remote_attestation/rust/src/tests/message.rs

```diff
@@ -0,0 +1,222 @@
+        SIGNING_ALGORITHM_KEY_LENGTH,
+    },
+    message::{
+        deserialize_message, ClientHello, ClientIdentity, Deserializable, EncryptedData,
+        MessageWrapper, Serializable, ServerIdentity, MAXIMUM_MESSAGE_SIZE,
+        REPLAY_PROTECTION_ARRAY_LENGTH,
+    },
+};
+use anyhow::{anyhow, Context};
+use assert_matches::assert_matches;
+use quickcheck::{quickcheck, TestResult};
+
+const INVALID_MESSAGE_HEADER: u8 = 5;
+
+/// Creates a zero initialized array.
+fn default_array<T, const L: usize>() -> [T; L]
+where
+    T: std::marker::Copy + std::default::Default,
+{
+    [Default::default(); L]
+}
+
+/// Converts slices to arrays (expands with zeroes).
+fn to_array<T, const L: usize>(input: &[T]) -> anyhow::Result<[T; L]>
+where
+    T: std::marker::Copy + std::default::Default,
+{
+    if input.len() <= L {
+        let mut result: [T; L] = default_array();
+        result[..input.len()].copy_from_slice(&input[..input.len()]);
+        Ok(result)
+    } else {
+        Err(anyhow!(
+            "Maximum input length exceeded, maximum {}, found {}",
+            L,
+            input.len()
+        ))
+    }
+}
+
+fn test_serialize_template<M: Serializable + Deserializable + PartialEq>(
+    message: &M,
+) -> anyhow::Result<bool> {
+    let serialized_message = message.serialize().context("Couldn't serialize message")?;
+    let deserialized_message =
+        M::deserialize(&serialized_message).context("Couldn't deserialize message")?;
+    Ok(*message == deserialized_message)
+}
+
+#[test]
+fn test_serialize_client_hello() {
+    fn property(random: Vec<u8>) -> TestResult {
+        if random.len() > REPLAY_PROTECTION_ARRAY_LENGTH {
+            return TestResult::discard();
+        }
+        let client_hello = ClientHello::new(to_array(&random).unwrap());
+        let result = test_serialize_template(&client_hello);
+        assert!(result.is_ok());
+        TestResult::from_bool(result.unwrap())
+    }
+    quickcheck(property as fn(Vec<u8>) -> TestResult);
+}
+
+#[test]
+fn test_serialize_server_identity() {
+    fn property(
+        ephemeral_public_key: Vec<u8>,
+        random: Vec<u8>,
+        transcript_signature: Vec<u8>,
+        signing_public_key: Vec<u8>,
+        attestation_info: Vec<u8>,
+    ) -> TestResult {
+        if ephemeral_public_key.len() > KEY_AGREEMENT_ALGORITHM_KEY_LENGTH
+            || random.len() > REPLAY_PROTECTION_ARRAY_LENGTH
+            || transcript_signature.len() > SIGNATURE_LENGTH
+            || signing_public_key.len() > SIGNING_ALGORITHM_KEY_LENGTH
+        {
+            return TestResult::discard();
+        }
+        let mut server_identity = ServerIdentity::new(
+            to_array(&ephemeral_public_key).unwrap(),
+            to_array(&random).unwrap(),
+            to_array(&signing_public_key).unwrap(),
+            attestation_info,
+        );
+        assert_eq!(server_identity.transcript_signature, default_array(),);
```

Comment:
What does quickcheck do when these assertions fail?

Reply:
It shows the test as failed, and tries to minimize the example input that made this test fail.